### PR TITLE
Fix product duplication issue on the Square side when WooCommerce is SOR.

### LINKED ISSUE #108: Duplicates in the Square catalog when hitting a rate limiting error
### Describe the bug

During catalog syncing, duplicates are created in the Square catalog when hitting a rate limiting error on Square's end.

Square's support insights on the duplicates:

> It does look like there is something off with these requests - the rate limiting error coming from us is not entirely correct, as some portion of the batch upsert request is successfully completing. This is why duplicates are being created; following the RATE_LIMITED error WooCommerce believes the entire request was not successful and retries it but some of items were already successfully created.

### To reproduce
To be replicated on our end, but this is what the merchant is doing is having a big set of products sync with Square. Enough to hit the rate limit.

### Expected behavior

If a rate limiting error is reached, there must not be duplicates when resuming the sync.

### Additional details

From Nic Wilson:

> The rate limit belongs to the Square account, which can be doing things other than interacting with WooCommerce. In this case, the account is also syncing the catalogue to Square Online. This means the limit is not something we can manage, so is not really our issue. Square acknowledge that.
However, the pickup and retry after the limit is applied is creating duplicate objects. This appears to be because the entire batch is rerun, even though half the batch may have been successful before failure. My guess is reusing the idempotency key helps prevent duplication of catalogue items.

Related idempotency key issue https://github.com/woocommerce/woocommerce-square/issues/106

8051230-zd-a8c
p1713482527612829-slack-C7U3Y3VMY

### LINKED ISSUE #232: Duplicates on upload/sync
### Issue

In 8942863-zen, On upload/sync, we get duplicate products created, and these duplicates correspond to errors such as:

`Product #3471 is excluded from sync as the product type "variation" is unsupported.`

This results in a stub product in Square where 3471 and it's SKU is showing up in what I would call an orphaned state, , as well as a properly synced product, which contains all the variations, #3471 and #3472 etc.

429 rate limiting causes duplicates #108 but we have reset the batch sizes, and rate limiting errors are no longer occurring.

SOR - WooCommerce
Square has a clean DB

The merchant tried cleaning products from Square and re-syncing, only to see more duplicates.

p1730380630885499-slack-C01TYRCJ1QD

### LINKED ISSUE #247: List Locations API is called on every visit to the Square sync update page
While working on #232, I noticed that the `listLocations` API was being called multiple times. As the plugin refreshes the page to update the sync status on each visit, the `listLocations` API gets triggered repeatedly. The `listLocations` API should only be called on the sync settings page, not on the sync status page.

### PR DESCRIPTION
**Pull request by iamdharmesh**
### All Submissions:

* [x] Does your code follow the [WooCommerce Sniffs](https://github.com/woocommerce/woocommerce-sniffs/) variant of WordPress coding standards?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?
* [ ] Will this change require new documentation or changes to existing documentation?

---

### Changes proposed in this Pull Request:
As reported in #232, when uploading or syncing new products with Square, issues arise when there are a large number of products to sync. The product sync fails, and duplicate products are created on the Square side. This occurred because the upsert limit was too high (500) for new products. As a result, the sync process experiences `429 Too Many Requests` errors sometimes. Additionally, in cases of successful requests, time and memory limitations prevent the Square item data from being stored in the product meta, causing the same product to be sent again for upsert, which leads to duplicate products. (It’s worth noting that Square allows the creation of products with the same SKU, which is unusual.)

This PR fixes the upsert process here and reduces the upsert limit from `500` to `25`, which enables the plugin to complete the upsert requests and update product data based on the response.

Closes #232.
Closes #247.
Closes #108.

### Steps to test the changes in this Pull Request:
1. Configure the plugin and set WooCommerce as the Source of Record (SOR).  
2. Create a large number of dummy products (2,000–3,000) using [WooCommerce Smooth Generator](https://github.com/woocommerce/wc-smooth-generator) or any other dummy data generation tool or just import products using CSV.
3. Enable sync for all products. (This could be done by passing "yes" under the "Sync with Square" column in CSV import in step 2)
4. Start a manual sync from the Square settings.  
5. Verify that the products are properly synced with Square and that no duplicate products are created on the Square side.

**#108**
1. Try sending too many requests to Square by following the steps [here](https://github.com/woocommerce/woocommerce-square-private/pull/1081#issue-1800438239).
1. Sending too many requests will result in 429 errors in sync. 
1. Verify no duplicate products are created on the Square side.

**#247**
1. Enable debug logs from Square settings.
2. Checkout to `trunk` branch
3. Visit `WooCommerce > Settings > Square > Update` page.
4. Notice there are logs for the `listLocations` API call.
5. Checkout to this branch.
6. Visit `WooCommerce > Settings > Square > Update` page.
7. Verify there is no log for the `listLocations` API call.

### CODE REVIEW

**Review by dkotter — APPROVED** [review]

**Comment by iamdharmesh** [comment]
@dkotter, I’ve implemented the retry for the 429 on this PR. Can you please review it quickly? Apologies for the re-review request. Thanks.

**Review by dkotter — APPROVED** [review]

**Comment by qasumitbagthariya** [comment]
### QA Update  ✅ 
-------------

I have verified this PR in the `fix/232` branch which has been fixed and is functioning as intended.

I tested the following on this branch:

I have cross-checked syncing 3000 products between Square and WooCommerce. No duplicate products were created on the Square side. Additionally, I verified that there are no logs for the listLocations API call. ✅

Closes https://github.com/woocommerce/woocommerce-square/issues/232.
Closes https://github.com/woocommerce/woocommerce-square/issues/247.
Closes https://github.com/woocommerce/woocommerce-square/issues/108.


**Testing Environment** 

<Details>

- WordPress: 6.7.1
- Theme: Twenty Twenty-Four 1.3
- WooCommerce - 9.4.1
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS Ventura 13.3 
- Branch: fix/232
</Details>


Steps to Test- As mentioned in the PR description.
Test Results - It is working as expected.
Functional Demo / Screencast - 
Special Notes -  Ready for code review (Woo)
Testing Document status:
Cases related to this Issue/PR are added to the Critical Flow Wiki pages: 

- [x] Yes
- [ ] Not Required/Applicable for this PR
------------------------------------------------

**Review by qasumitbagthariya — APPROVED** [review]

**Comment by vikrampm1** [comment]
@diegocurbelo this PR is ready for your review!

File: includes/Settings.php [hunk 1]
```diff
@@ -789,7 +789,7 @@ public function get_locations( $force = false ) {
 
 		$section = isset( $_GET['section'] ) ? sanitize_text_field( wp_unslash( $_GET['section'] ) ) : false;  // phpcs:ignore WordPress.Security.NonceVerification.Recommended
 
-		if ( ! ( $this->is_admin_settings_screen() || ( $this->is_admin_settings_screen() && 'update' !== $section ) || $force ) ) {
+		if ( ! ( $this->is_admin_settings_screen() && ( $this->is_admin_settings_screen() && 'update' !== $section ) || $force ) ) {
```
Comment (diegocurbelo) [inline includes/Settings.php]: `$this->is_admin_settings_screen()` is checked twice:

```suggestion
		if ( ! (  ( $this->is_admin_settings_screen() && 'update' !== $section ) || $force ) ) {
```

**Review by diegocurbelo — APPROVED** [review]
Thanks for working on this @iamdharmesh!

The code cahnges looks good, and it tests well.
I left a code suggestion, but it has no functional impact, it's for readability.

**Review by jeffpaul — APPROVED** [review]

**Comment by qasumitbagthariya** [comment]
Regression / Smoke Test Report  ✅ 
-------

Tested with Archive File created via "php woorelease.phar build repo_URL" (Composer version 2.5.5, npm version 8.19.4, node version 16.20.0)

Status- Working expected with Plugin Archive/Zip file same as fix specific branch.

**Testing Environment** 

<Details>

- WordPress: 6.7.1
- Theme: Twenty Twenty-Four 1.3
- WooCommerce - 9.4.1
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS Ventura 13.3 
- Branch: smoke-testing

</Details>


Next Step- Ready to Merge 🚀

status update:
@vikrampm1

**Pull request merged**